components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/CustomUserSelfRegistrationHandler.java

```diff
@@ -0,0 +1,101 @@
+
+import java.util.List;
+
+import static java.util.Arrays.asList;
+
+public class CustomUserSelfRegistrationHandler extends AbstractEventHandler {
+
+    //Role Constants
+    public static final String SUBSCRIBER_ROLE = "Internal/subscriber";
+    public static final String SELF_SIGNUP_ROLE = "Internal/selfsignup";
+
+
+    @Override public String getName() {
+        return "customUserSelfRegistration";
+    }
+
+    public void handleEvent(Event event) throws IdentityEventException {
+
+        String tenantDomain = (String) event.getEventProperties()
+                .get(IdentityEventConstants.EventProperty.TENANT_DOMAIN);
+        String userName = (String) event.getEventProperties().get(IdentityEventConstants.EventProperty.USER_NAME);
+
+        //The handler should be called ss a post add user event.
+        if (IdentityEventConstants.Event.POST_ADD_USER.equals(event.getEventName())) {
+            try {
+                addNewRole(tenantDomain, userName);
+            } catch (IdentityRecoveryServerException e) {
+                throw new IdentityEventException("Error while adding custom roles to the user", e);
+            }
+        }
+    }
+
+    private void addNewRole(String tenantDomain, String userName)
+            throws org.wso2.carbon.identity.recovery.IdentityRecoveryServerException {
+        try {
+            //Realm service is used for user management tasks
+            RealmService realmService = APIKeyMgtDataHolder.getRealmService();
+            UserStoreManager userStoreManager;
+            try {
+                userStoreManager = realmService.getTenantUserRealm(IdentityTenantUtil.getTenantId(tenantDomain))
+                        .getUserStoreManager();
+            } catch (UserStoreException e) {
+                throw Utils
+                        .handleServerException(IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_UNEXPECTED, userName,
+                                e);
+            }
+            //Start a tenant flow
+            PrivilegedCarbonContext.startTenantFlow();
+            PrivilegedCarbonContext carbonContext = PrivilegedCarbonContext.getThreadLocalCarbonContext();
+            carbonContext.setTenantId(IdentityTenantUtil.getTenantId(tenantDomain));
+            carbonContext.setTenantDomain(tenantDomain);
+            try {
+                //Since this handler is called as a post add user event, the user should exists in the userstore
+                if (userStoreManager.isExistingUser(userName)) {
+                    List<String> roleList = asList(userStoreManager.getRoleListOfUser(userName));
+                    //User should have selfSignup role. Checking whether the user is in the new role
+                    if (roleList.contains(SELF_SIGNUP_ROLE) && !roleList.contains(SUBSCRIBER_ROLE)) {
+                        String[] userRoles = new String[] { SUBSCRIBER_ROLE };
+                        userStoreManager.updateRoleListOfUser(userName, null, userRoles);
```

Comment:
IIRC updateRoleListOfUser method removes the existing roles and add new roles given in the userRoles.
Please check whether SELF_SIGNUP_ROLE is getting removed in this case.

Reply:
Internal/selfsignup role doesn't get removed from the roles, only internal/subscriber will be added as a new entry to the existing list.